tests/sentry/backup/test_correctness.py

```diff
@@ -0,0 +1,164 @@
+        )
+
+    def __hash__(self):
+        return hash((self.model, self.pk))
+
+    def print(self):
+        return f'InstanceID(model: "{self.model}", pk: {self.pk})'
+
+
+class ComparatorFinding:
+    """Store all information about a single failed matching between expected and actual output."""
+
+    def __init__(self, name: ComparatorName, on: InstanceID, reason: str | None = ""):
+        self.name = name
+        self.on = on
+        self.reason = reason
+
+    def print(self):
+        return f'Finding(\n\tname: "{self.name}",\n\ton: {self.on.print()},\n\treason: {self.reason}\n)'
+
+
+class ComparatorFindings:
+    """A wrapper type for a list of 'ComparatorFinding' which enables pretty-printing in asserts."""
+
+    def __init__(self, findings: list[ComparatorFinding]):
+        self.findings = findings
+
+    def append(self, finding: ComparatorFinding):
+        self.findings.append(finding)
+
+    def assert_on_findings(self):
+        if self.findings:
+            assert False, self.print()
+
+    def print(self):
+        return "\n".join(map(lambda f: f.print(), self.findings))
+
+
+REPO_PATH = Path(os.path.dirname(os.path.realpath("__file__")))
+FIXTURE_PATH = REPO_PATH / "fixtures/backup"
+INDENT = 2
+JSON_PRETTY_PRINTER = JSONEncoder(
+    default=better_default_encoder, indent=INDENT, ignore_nan=True, sort_keys=True
+)
+
+
+def json_lines(obj: JSONData) -> list[str]:
+    """Take a JSONData object and pretty-print it as JSON."""
+
+    return JSON_PRETTY_PRINTER.encode(obj).splitlines()
+
+
+# TODO(team-ospo/#155): Move this out of the test suite, and into its own standalone module, since eventually it will be used to compare live JSON as well.
+def validate(expect: JSONData, actual: JSONData) -> ComparatorFindings:
+    """Ensures that originally imported data correctly matches actual outputted data, and produces a list of reasons why not when it doesn't"""
+
+    findings = ComparatorFindings([])
+    exp_models = {}
+    act_models = {}
+    for model in expect:
+        id = InstanceID(ModelName(model["model"]), PositiveInt(model["pk"]))
+        exp_models[id] = model
+
+    # Ensure that the actual JSON contains no duplicates - we assume that the expected JSON did not.
+    for model in actual:
+        id = InstanceID(ModelName(model["model"]), PositiveInt(model["pk"]))
+        if id in act_models:
+            findings.append(ComparatorFinding(ComparatorName("duplicate_entry"), id))
+        else:
+            act_models[id] = model
+
+    # Report unexpected and missing entries in the actual JSON.
+    act_ids = list(act_models)
+    exp_ids = list(exp_models)
+    for id in act_ids:
+        if id not in exp_ids:
+            del act_models[id]
+            findings.append(ComparatorFinding(ComparatorName("unexpected_entry"), id))
+    for id in exp_ids:
+        if id not in act_ids:
+            del exp_models[id]
+            findings.append(ComparatorFinding(ComparatorName("missing_entry"), id))
```

Comment:
this would be better done with set math:

```python
missing = sorted(exp_models.keys() - act_models.keys())
extra = sorted(act_models.keys() - exp_models.keys())
```

Reply:
Done